### PR TITLE
Submit new JWKs to consensus and commit them to the chain

### PR DESCRIPTION
**Pull request by mystenmark**
## Description 

First part of tracking JWKs via consensus. 

Outline of PR:
- every node fetches JWKs periodically, submits them to consensus as NewJWKFetched messages
- consensus_handler counts the votes, and adds AuthenticatorStateUpdate transactions to checkpoints.
- After execution, the node adds the new JWKs to the local signature verifier so that they can be used for verifying transactions.

There is still some work remaining to support reconfiguration, that will be done in a subsequent PR.

## Test Plan 

Tests coming soon

---

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

**Comment by lxfind** [comment]
Have you tried to create the new object locally yet? One nit is that the runtime may treat it as unwrapped instead of created due to the ID not recorded

File: crates/sui-json-rpc-types/src/sui_transaction.rs [hunk 1]
```diff
@@ -1037,6 +1046,10 @@ pub struct SuiConsensusCommitPrologue {
     pub commit_timestamp_ms: u64,
 }
 
+#[serde_as]
+#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
+pub struct SuiAuthenticatorStateUpdate {}
```
Comment (andll) [inline crates/sui-json-rpc-types/src/sui_transaction.rs:1051]: Looks like we made a choice not to expose data from this txn to jsonrpc - is this because clients can't do much with this data?
Even so, I would still expose the data just for uniformity with other RPCs, unless there is other reason to hide this data

File: crates/sui-node/src/lib.rs [hunk 2]
```diff
@@ -200,68 +202,87 @@ impl SuiNode {
...
+
+            let mut join_set = tokio::task::JoinSet::new();
+            for p in supported_providers.into_iter() {
+                let epoch_store = epoch_store.clone();
+                let consensus_adapter = consensus_adapter.clone();
+                join_set.spawn(
+                    async move {
+                        // note: restart-safe de-duplication happens after consensus, this is
+                        // just best-effort to reduce unneeded submissions.
+                        let mut seen = HashSet::new();
```
Comment (andll) [inline crates/sui-node/src/lib.rs:231]: Should we at least load currently active JWKs? Otherwise it seems every time node restart it will re-send all jwks to consensus?

File: crates/sui-node/src/lib.rs [hunk 3]
```diff
@@ -200,68 +202,87 @@ impl SuiNode {
...
                     }
-                }
-                info!("JWK updater task terminated");
+                    .instrument(error_span!("jwk_updater_task", epoch)),
+                );
             }
-            .instrument(error_span!("jwk_updater_task", epoch)),
-        );
+
+            epoch_store.wait_epoch_terminated().await;
```
Comment (andll) [inline crates/sui-node/src/lib.rs:268]: Some nit picking on the tasks - right now we send stop signal to task when epoch ends, but tasks can still continue running while epoch end happens, as we don't wait for the to stop.
This is probably ~ok, but perhaps we should just add a "barrier" here, to make sure those tasks stop before epoch ends, just so that we avoid questions like "can something go wrong".

We have a convenient function for this - `AuthorityEpochStore::within_alive_epoch` - it drops task when epoch ends and waits for it to complete before proceeding with reconfiguration

File: crates/sui-types/src/authenticator_state.rs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    pub id: UID,
+    pub version: u64,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct AuthenticatorStateInner {
+    pub version: u64,
+
+    /// List of currently active JWKs.
+    pub active_jwks: Vec<ActiveJwk>,
```
Comment (andll) [inline crates/sui-types/src/authenticator_state.rs:32]: Do we have a limit on this size? Any chance now/in the future we could come close to this limit?

File: crates/sui-types/src/messages_consensus.rs [hunk 5]
```diff
@@ -34,12 +48,15 @@ pub struct ConsensusTransaction {
...
 }
 
-#[derive(Serialize, Deserialize, Clone, Copy, Hash, PartialEq, Eq)]
+#[derive(Serialize, Deserialize, Clone, Hash, PartialEq, Eq)]
 pub enum ConsensusTransactionKey {
     Certificate(TransactionDigest),
     CheckpointSignature(AuthorityName, CheckpointSequenceNumber),
     EndOfPublish(AuthorityName),
     CapabilityNotification(AuthorityName, u64 /* generation */),
+    // Key must include both id and jwk, because honest validators could be given multiple jwks for
```
Comment (andll) [inline crates/sui-types/src/messages_consensus.rs:57]: Btw, I wonder if this could have been prevented, if we chose JwkId = Hash(JWK)? Is there reason we chose a schema that allows multiple JWKs with same id?

**Review by andll — COMMENTED** [review]
Added some comments, some of them (like JWK id question) are probably from my misunderstanding of how JWK works...

File: crates/sui-node/src/lib.rs (hunk 3 above)
Comment (mystenmark) [reply crates/sui-node/src/lib.rs:268]: good idea, will add that. I can probably get rid of the join set then

File: crates/sui-types/src/messages_consensus.rs (hunk 5 above)
Comment (mystenmark) [reply crates/sui-types/src/messages_consensus.rs:57]: If we were generating the data ourselves, we could do something like that. However, a malfunctioning oauth provider could actually return conflicting jwks. Similarly, the same JWK could be named by more than one ID, and a hash wouldn't handle that case correctly. The first case will probably never happen, but the second case seems at least conceivable to me and I'll sleep better knowing we handle it correctly.

File: crates/sui-node/src/lib.rs (hunk 2 above)
Comment (mystenmark) [reply crates/sui-node/src/lib.rs:231]: That shouldn't be very many messages, but it will be easy enough to handle this.

File: crates/sui-core/src/authority/authority_per_epoch_store.rs [hunk 6]
```diff
@@ -183,6 +188,9 @@ pub struct AuthorityPerEpochStore {
 
     /// Chain identifier
     chain_identifier: ChainIdentifier,
+
+    /// aggregator for JWK votes
+    jwk_aggregator: OnceCell<Arc<Mutex<JwkAggregator>>>,
```
[the remark this answers is not in the document]
Comment (mystenmark) [reply crates/sui-core/src/authority/authority_per_epoch_store.rs:193]: I was thinking we would simply put a limit on the number of votes each validator can send per epoch - wdyt?

File: crates/sui-core/src/authority/authority_per_epoch_store.rs [hunk 7]
```diff
@@ -1217,6 +1229,73 @@ impl AuthorityPerEpochStore {
...
+            .iter_with_bounds(Some(round), Some(round))
+            .map(|(r, (jwk_id, jwk))| {
+                debug_assert!(round == r);
+                ActiveJwk { jwk_id, jwk, epoch }
+            })
+            .collect())
+    }
+
+    fn get_or_recover_jwk_aggregator(&self) -> Arc<Mutex<JwkAggregator>> {
+        self.jwk_aggregator
```
[the remark this answers is not in the document]
Comment (mystenmark) [reply crates/sui-core/src/authority/authority_per_epoch_store.rs:1281]: I think it made more sense in an earlier version of the code but there's no reason to do it any more

File: crates/sui-core/src/authority.rs [hunk 8]
```diff
@@ -1224,15 +1231,11 @@ impl AuthorityState {
                 error: "dry-exec is only supported on fullnodes".to_string(),
             });
         }
-        match transaction.kind() {
-            TransactionKind::ProgrammableTransaction(_) => (),
```
[the remark this answers is not in the document]
Comment (mystenmark) [reply crates/sui-core/src/authority.rs:1228]: previous code: if PT, pass, otherwise return error (by enumerating all cases)

new behavior: if system_tx, return error, otherwise pass (because it is implied to be a PT).

File: crates/sui-core/src/transaction_input_checker.rs [hunk 9]
```diff
@@ -121,17 +124,12 @@ mod checked {
     ) -> SuiResult<(ObjectRef, InputObjects)> {
         let gas_object_ref = gas_object.compute_object_reference();
         kind.validity_check(config)?;
-        match kind {
-            TransactionKind::ProgrammableTransaction(_) => (),
```
[the remark this answers is not in the document]
Comment (mystenmark) [reply crates/sui-core/src/transaction_input_checker.rs:125]: same answer

File: crates/sui-types/src/authenticator_state.rs (hunk 4 above)
Comment (mystenmark) [reply crates/sui-types/src/authenticator_state.rs:32]: There is currently no limit on how many we store on chain. The fetch side is limited to 100 JWKs per provider per hour.

Suppose that one oauth provider goes rogue and tries to inundate us with bogus keys. That would be 100 new keys every hour. Google keys are about 500 bytes each. Eventually we will only keep keys from the current and prior epoch, so we would have about 2.5MB of keys in storage.

Of course, they could also send us extremely large keys. I'm adding code now to limit the size of ID + key to 4096 bytes.

File: crates/sui-json-rpc-types/src/sui_transaction.rs (hunk 1 above)
Comment (mystenmark) [reply crates/sui-json-rpc-types/src/sui_transaction.rs]: done

**Comment by mystenmark** [comment]
@andll thanks for the super helpful comments!

File: crates/sui-core/src/authority/authority_per_epoch_store.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (mystenmark) [reply crates/sui-core/src/authority/authority_per_epoch_store.rs]: I added a vote limit per epoch in the code

[the remark this answers is not in the document]
Comment (andll) [reply crates/sui-core/src/authority/authority_per_epoch_store.rs]: Yeah I think this makes sense

File: crates/sui-core/src/authority.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (andll) [reply crates/sui-core/src/authority.rs:1228]: Ok, got it

**Review by andll — APPROVED** [review]
Looks like remaining parts are to call move code to update on-chain jwks, and figure out something for the reconfiguration? (assuming both of those for the future PRs?)

I also have a question about narwhal, i will ask it to narwhal channel to loop in more folks, but I think this large PR should probably be ok to merge

File: crates/sui-node/src/lib.rs [hunk 10]
```diff
@@ -200,68 +204,84 @@ impl SuiNode {
...
+                    // note: restart-safe de-duplication happens after consensus, this is
+                    // just best-effort to reduce unneeded submissions.
+                    let mut seen = HashSet::new();
+                    loop {
+                        info!("fetching JWK for provider {:?}", p);
+                        match Self::fetch_jwks(&p).await {
                             Err(e) => {
                                 warn!("Error when fetching JWK {:?}", e);
                                 // Retry in 30 seconds
                                 tokio::time::sleep(Duration::from_secs(30)).await;
```
Comment (akichidis) [inline crates/sui-node/src/lib.rs:238]: Not part of this PR, but I believe in a case of error this will wait here for 30 sec, but then the next execution [line will be 265](https://github.com/MystenLabs/sui/pull/13530/files#diff-3fb10736526598f38da2b909064220eb69d70debb261372b7f452525105aa1a2R265) where we wait again for 1 hour - based on the comment I understand that we wanted to re-fetch again the jwks within 30 secs, right ?

Comment (mystenmark) [reply crates/sui-node/src/lib.rs:238]: Good catch, thanks!

File: sui-execution/latest/sui-adapter/src/execution_engine.rs [hunk 11]
```diff
@@ -502,6 +502,9 @@ mod checked {
                     pt,
                 )
             }
+            TransactionKind::AuthenticatorStateUpdate(_) => {
+                todo!()
```
[the remark this answers is not in the document]
Comment (mystenmark) [reply sui-execution/latest/sui-adapter/src/execution_engine.rs:506]: Correct - this code will not run yet, the implementation here is coming in the next PR
